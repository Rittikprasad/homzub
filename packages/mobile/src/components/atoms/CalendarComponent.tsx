import React, { Component } from 'react';
import { Calendar, DateObject } from 'react-native-calendars';
import { View, FlatList, TouchableOpacity, StyleSheet, StyleProp, TextStyle, ViewStyle } from 'react-native';
import moment from 'moment';
import { groupBy } from 'lodash';
import { DateFormats, DateUtils, MonthNames } from '@homzhub/common/src/utils/DateUtils';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import CalendarHeader from '@homzhub/common/src/components/atoms/CalendarHeader';
import { Label } from '@homzhub/common/src/components/atoms/Text';

// CONSTANTS START

const INITIAL_YEAR = 1893;
const MAX_YEAR = Number(DateUtils.getNextYear(16));

// CONSTANTS END

interface ICalendarProps {
  onSelect: (day: string) => void;
  selectedDate: string;
  isCurrentDateEnable?: boolean;
  allowPastDates?: boolean;
  maxDate?: string;
  minDate?: string;
  isOnlyYearView?: boolean;
}

interface ICalendarState {
  isMonthView: boolean;
  isYearView: boolean;
  selectedDate: string;
  day: string;
  month: number;
  year: string;
  yearTitle: string;
  yearList: string[];
}

export class CalendarComponent extends Component<ICalendarProps, ICalendarState> {
  public constructor(props: ICalendarProps) {
    super(props);
    const { selectedDate, isCurrentDateEnable } = props;
    this.state = {
      isMonthView: false,
      isYearView: false,
      selectedDate: isCurrentDateEnable ? DateUtils.getCurrentDate() : selectedDate,
      day: moment().format(DateFormats.DD),
      month: new Date().getMonth(),
      year: moment().format(DateFormats.YYYY),
      yearTitle: '',
      yearList: [],
    };
  }

  public componentDidMount = (): void => {
    const { selectedDate, isOnlyYearView } = this.props;
    let yearData = this.getYearsData();
    if (isOnlyYearView) {
      yearData = this.getYearsData(Number(selectedDate));
    }
    const title = `${yearData[0]} - ${yearData[yearData.length - 1]}`;
    this.setState({ yearTitle: title, yearList: yearData });
  };

  public render(): React.ReactNode {
    const { isMonthView, isYearView } = this.state;
    const { isOnlyYearView } = this.props;
    const isYear = isYearView || isOnlyYearView;
    return (
      <View>
        {this.renderHeader()}
        {isMonthView ? this.renderListView() : isYear ? this.renderListView(true) : this.renderCalendar()}
      </View>
    );
  }

  private renderHeader = (): React.ReactElement => {
    const { allowPastDates, maxDate, isOnlyYearView, minDate } = this.props;
    const { isMonthView, selectedDate, isYearView, yearTitle } = this.state;
    let { month, year } = this.state;
    if (selectedDate) {
      month = moment(selectedDate).month();
      year = String(moment(selectedDate).year());
    }
    const updateMonth = DateUtils.getFullMonthName(month, DateFormats.MMMM);

    const isCurrentMonth = month === moment().month();

    const title = `${updateMonth} ${year}`;

    return (
      <CalendarHeader
        isAllowPastDate={allowPastDates}
        headerTitle={title}
        month={month}
        year={year}
        headerYear={`${year}`}
        isCurrentMonth={isCurrentMonth}
        isMonthView={isMonthView}
        yearTitle={yearTitle}
        isYearView={isYearView || isOnlyYearView}
        maxDate={maxDate}
        minDate={minDate}
        onBackPress={this.handleBackPress}
        onNextPress={this.handleNextPress}
        onMonthPress={this.handleMonthPress}
        onYearPress={this.handleYearPress}
      />
    );
  };

  private renderListView = (isYearView?: boolean): React.ReactNode => {
    const { yearList } = this.state;

    const listData = isYearView ? yearList : MonthNames;

    return (
      <FlatList
        data={listData}
        renderItem={this.renderItem}
        contentContainerStyle={styles.listContent}
        keyExtractor={this.renderKeyExtractor}
        numColumns={4}
      />
    );
  };

  private renderKeyExtractor = (item: string, index: number): string => {
    return `${item}-${index}`;
  };

  private renderItem = ({ item, index }: { item: string; index: number }): React.ReactElement => {
    const { month, isMonthView, isYearView, year } = this.state;
    const { isOnlyYearView, selectedDate } = this.props;
    const onPressItem = (): void => (isMonthView ? this.onSelectMonth(item, index) : this.onSelectYear(item, index));
    const yearView = isOnlyYearView ? Number(selectedDate) === Number(item) : month === index;
    const isSelected = isYearView ? year === item : yearView;
    const isMobile = PlatformUtils.isMobile();

    return (
      <TouchableOpacity
        key={index}
        style={StyleSheet.flatten([customStyles.renderItemView(isSelected, isMobile)])}
        onPress={onPressItem}
      >
        <Label type="large" style={StyleSheet.flatten([customStyles.renderItemTitle(isSelected)])}>
          {item}
        </Label>
      </TouchableOpacity>
    );
  };

  private renderCalendar = (): React.ReactElement => {
    const { allowPastDates, maxDate, minDate } = this.props;
    const { day, month, year, selectedDate } = this.state;
    const updateMonth = month + 1;
    const date = selectedDate || DateUtils.getFormattedDate(day, updateMonth, year, 'YYYY-MM-DD').toDateString();
    const markedDate = !selectedDate ? moment().format('YYYY-MM-DD') : moment(date).format('YYYY-MM-DD');

    return (
      <>
        <Calendar
          hideArrows
          // @ts-ignore
          renderHeader={(): null => null}
          minDate={minDate ? DateUtils.getDisplayDate(minDate, 'YYYY-MM-DD') : allowPastDates ? undefined : new Date()}
          maxDate={maxDate}
          current={date}
          key={date}
          style={styles.calendarStyle}
          onDayPress={this.onDayPress}
          markingType="custom"
          theme={{}}
          markedDates={{
            [markedDate]: {
              customStyles: {
                container: {
                  backgroundColor: theme.colors.primaryColor,
                  borderRadius: 4,
                },
                text: {
                  color: theme.colors.white,
                },
              },
            },
          }}
        />
        <Button
          type="primary"
          title="Select"
          containerStyle={styles.buttonStyle}
          disabled={!selectedDate}
          onPress={this.handleSelect}
        />
      </>
    );
  };

  private onSelectMonth = (item: string, index: number): void => {
    const { year } = this.state;
    this.setState({ month: index, isMonthView: false, isYearView: false });
    this.getSelectedDate(index, Number(year));
  };

  private onSelectYear = (item: string, index: number): void => {
    const { month } = this.state;
    const { isOnlyYearView, onSelect } = this.props;
    if (isOnlyYearView) {
      this.setState({ year: item, isMonthView: false, isYearView: false });
      onSelect(item);
    } else {
      this.setState({ year: item, isMonthView: true, isYearView: false });
      this.getSelectedDate(Number(month), index);
    }
  };

  private onDayPress = (day: DateObject): void => {
    this.setState({ selectedDate: day.dateString });
  };

  private handleSelect = (): void => {
    const { onSelect } = this.props;
    const { selectedDate } = this.state;
    onSelect(selectedDate);
  };

  private handleMonthPress = (): void => {
    const { isMonthView } = this.state;
    this.setState({ isMonthView: !isMonthView, isYearView: false });
  };

  private handleYearPress = (): void => {
    const { isYearView } = this.state;
    const yearData = this.getYearsData();
    const yearTitle = `${yearData[0]} - ${yearData[yearData.length - 1]}`;
    this.setState({ isYearView: !isYearView, isMonthView: false, yearTitle, yearList: yearData });
  };

  /**
   * Handle Back Press Functionality
   * Cases: Day, Month and Year view
   */
  private handleBackPress = (): void => {
    const { allowPastDates, isOnlyYearView } = this.props;
    const { isMonthView, selectedDate, isYearView, yearTitle } = this.state;
    let { month, year } = this.state;

    // For year view
    const value = Number(yearTitle.split('-')[0]);
    // For year view

    if (selectedDate) {
      month = moment(selectedDate).month();
      year = moment(selectedDate).year().toString();
    }

    if (!allowPastDates && month === moment().month() && !isYearView) {
      return;
    }

    if (isMonthView) {
      const previousYear = Number(year) > INITIAL_YEAR ? Number(year) - 1 : Number(year);
      this.getSelectedDate(month, previousYear);
      this.setState({ year: previousYear.toString() });
    } else if ((isYearView || isOnlyYearView) && value > INITIAL_YEAR) {
      const previousYear = value - 1;
      const yearData = this.getYearsData(previousYear);
      const title = `${yearData[0]} - ${yearData[yearData.length - 1]}`;
      this.setState({ yearTitle: title, yearList: yearData });
    } else {
      const previousMonth = month < 1 ? 11 : month - 1;
      const updatedYear = month < 1 ? Number(year) - 1 : Number(year);
      this.getSelectedDate(previousMonth, updatedYear);
      this.setState({ month: previousMonth, year: updatedYear.toString() });
    }
  };

  /**
   * Handle Next Press Functionality
   * Cases: Day, Month and Year view
   */
  private handleNextPress = (): void => {
    const { isMonthView, selectedDate, isYearView, yearTitle } = this.state;
    const { maxDate, isOnlyYearView } = this.props;

    // For year view
    const value = Number(yearTitle.split('-')[1]);
    // For year view

    let { month, year } = this.state;
    if (maxDate && moment(maxDate).month() === moment().month()) {
      return;
    }
    if (selectedDate) {
      month = moment(selectedDate).month();
      year = moment(selectedDate).year().toString();
    }

    if (isMonthView) {
      const nextYear = Number(year) < MAX_YEAR - 1 ? Number(year) + 1 : Number(year);
      this.getSelectedDate(month, nextYear);
      this.setState({ year: nextYear.toString() });
    } else if ((isYearView || isOnlyYearView) && value < MAX_YEAR - 1) {
      const nextYear = value + 1;
      const yearData = this.getYearsData(nextYear);
      const title = `${yearData[0]} - ${yearData[yearData.length - 1]}`;
      this.setState({ yearTitle: title, yearList: yearData });
    } else {
      const nextMonth = month <= 10 ? month + 1 : 0;
      const nextYear = month <= 10 ? Number(year) : Number(year) + 1;
      this.getSelectedDate(nextMonth, nextYear);
      this.setState({ month: nextMonth, year: nextYear.toString() });
    }
  };

  private getSelectedDate = (month: number, year: number): void => {
    const { selectedDate } = this.props;
    const isNotSelected = moment(selectedDate).month() !== month || moment(selectedDate).year() !== year;
    if (isNotSelected) {
      this.setState({ selectedDate: '' });
    } else {
      this.setState({ selectedDate });
    }
  };

  /**
   * Created Year list for year view
   * Grouping by INITIAL and MAX year on the gap of 16 years
   */
  private getYearsData = (updateYear?: number): string[] => {
    const { year, selectedDate } = this.state;
    const newYear = moment(selectedDate).year();
    const formattedYear = updateYear || Number(newYear || year);
    const years = [];
    let updatedData: string[] = [];

    // for years list
    for (let i = INITIAL_YEAR; i <= MAX_YEAR; i++) {
      if (i < MAX_YEAR) {
        years.push(i.toString());
      }
    }

    while (years.length > 0) {
      const data = years.splice(0, 16);
      const abc = groupBy(data, () => `${data[0]}-${data[data.length - 1]}`);

      // eslint-disable-next-line no-loop-func
      Object.keys(abc).forEach((key) => {
        const values: string[] = key.split('-');
        if (Number(values[0]) <= formattedYear && Number(values[1]) >= formattedYear) {
          updatedData = abc[key];
        }
      });
    }

    return updatedData;
  };
}

const styles = StyleSheet.create({
  buttonStyle: {
    flex: 0,
    margin: 16,
  },
  listContent: {
    marginVertical: 16,
  },
  calendarStyle: {
    height: 310,
  },
});

const customStyles = {
  headerTitle: (isMonthView: boolean): StyleProp<TextStyle> => ({
    color: isMonthView ? theme.colors.darkTint2 : theme.colors.primaryColor,
  }),
  renderItemView: (isSelected: boolean, isMobile: boolean): StyleProp<ViewStyle> => ({
    width: isMobile ? theme.viewport.width * 0.24 : 60,
    marginVertical: 12,
    alignItems: 'center',
    backgroundColor: isSelected ? theme.colors.primaryColor : theme.colors.white,
    borderRadius: 4,
  }),
  renderItemTitle: (isSelected: boolean, isDisable?: boolean): StyleProp<TextStyle> => ({
    paddingVertical: 6,
    color: isSelected ? theme.colors.white : isDisable ? theme.colors.disabled : theme.colors.darkTint2,
  }),
};
