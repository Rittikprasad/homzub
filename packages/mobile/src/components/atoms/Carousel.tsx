import React from 'react';
import Carousel from 'react-native-snap-carousel';
import { StyleProp, ViewStyle } from 'react-native';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { theme } from '@homzhub/common/src/styles/theme';

interface ICarouselProps<T> {
  carouselData: T[];
  carouselItem: (item: T) => React.ReactElement;
  activeIndex: number;
  onSnapToItem: (index: number) => void;
  sliderWidth?: number;
  initialNumToRender?: number;
  itemWidth?: number;
  bubbleRef?: (ref: any) => void;
  contentStyle?: StyleProp<ViewStyle>;
  containerStyle?: StyleProp<ViewStyle>;
  testID?: string;
  extraData?: number[];
  onLoadMore?: () => void;
  hasParallaxImages?: boolean;
}

const ACTIVE_SLIDE_OFFSET = 20;

export class SnapCarousel<T> extends React.PureComponent<ICarouselProps<T>> {
  private carouselRef: Carousel<T> | null = null;

  public render = (): React.ReactElement => {
    const {
      carouselData,
      activeIndex,
      containerStyle,
      contentStyle,
      testID,
      onSnapToItem,
      initialNumToRender = 10,
      sliderWidth = theme.viewport.width,
      itemWidth = theme.viewport.width - 30,
      extraData,
      onLoadMore,
      hasParallaxImages = false,
    } = this.props;

    return (
      // @ts-ignore
      <Carousel
        testID={testID}
        data={carouselData}
        firstItem={activeIndex}
        sliderWidth={sliderWidth}
        itemWidth={itemWidth}
        renderItem={this.renderItem}
        hasParallaxImages={hasParallaxImages}
        initialNumToRender={initialNumToRender}
        activeSlideOffset={ACTIVE_SLIDE_OFFSET}
        contentContainerCustomStyle={contentStyle}
        containerCustomStyle={containerStyle}
        onLayout={this.updateRef}
        onEndReached={({ distanceFromEnd }: { distanceFromEnd: number }): void => {
          if (distanceFromEnd > 0 && onLoadMore) {
            onLoadMore();
          }
        }}
        onSnapToItem={onSnapToItem}
        extraData={extraData}
        removeClippedSubviews={PlatformUtils.isAndroid()}
        ref={(c): void => {
          this.carouselRef = c;
        }}
      />
    );
  };

  public renderItem = ({ item }: { item: T }): React.ReactElement => {
    const { carouselItem } = this.props;
    return carouselItem(item);
  };

  public updateRef = (): void => {
    const { bubbleRef } = this.props;
    if (bubbleRef) {
      bubbleRef(this.carouselRef);
    }
  };
}
