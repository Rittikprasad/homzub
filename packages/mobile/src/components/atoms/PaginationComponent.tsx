import React from 'react';
import { Pagination } from 'react-native-snap-carousel';
import { StyleProp, StyleSheet, ViewStyle } from 'react-native';
import { theme } from '@homzhub/common/src/styles/theme';

interface IPaginationProps {
  dotsLength: number;
  activeSlide: number;
  containerStyle?: StyleProp<ViewStyle>;
  activeDotStyle?: StyleProp<ViewStyle>;
  inactiveDotStyle?: StyleProp<ViewStyle>;
}

export const PaginationComponent = (props: IPaginationProps): React.ReactElement => {
  const { dotsLength, activeSlide, containerStyle, inactiveDotStyle, activeDotStyle } = props;
  return (
    <Pagination
      dotsLength={dotsLength}
      activeDotIndex={activeSlide}
      dotStyle={[styles.dotStyle, activeDotStyle]}
      inactiveDotStyle={[styles.inactiveDotStyle, inactiveDotStyle]}
      inactiveDotOpacity={1}
      inactiveDotScale={1}
      containerStyle={containerStyle}
    />
  );
};

const styles = StyleSheet.create({
  dotStyle: {
    width: 10,
    height: 10,
    marginHorizontal: 2,
    borderStyle: 'solid',
    borderWidth: 3,
    borderRadius: 5,
    borderColor: theme.colors.blue,
    backgroundColor: theme.colors.white,
  },
  inactiveDotStyle: {
    backgroundColor: theme.colors.blue,
  },
});
