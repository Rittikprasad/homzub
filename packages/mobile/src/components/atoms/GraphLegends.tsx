import React from 'react';
import { View, StyleSheet, ScrollView } from 'react-native';
import { useSelector } from 'react-redux';
import { theme } from '@homzhub/common/src/styles/theme';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import { PricePerUnit } from '@homzhub/common/src/components/atoms/PricePerUnit';
import { IGeneralLedgerGraphData } from '@homzhub/common/src/domain/models/GeneralLedgers';

interface IProps {
  data: IGeneralLedgerGraphData[];
}

const GraphLegends = (props: IProps): React.ReactElement => {
  const { data } = props;
  const currency = useSelector(UserSelector.getCurrency);

  return (
    <ScrollView horizontal showsHorizontalScrollIndicator={false} contentContainerStyle={styles.container}>
      {data.map((legend: IGeneralLedgerGraphData, index: number) => (
        <View style={styles.legendContainer} key={`legend-${index}`}>
          <View style={[styles.color, { backgroundColor: legend.svg.fill }]} />
          <View>
            <Label type="regular" textType="regular" style={styles.textColor}>
              {legend.title}
            </Label>
            <PricePerUnit price={legend.value} currency={currency} textSizeType="small" textFontWeight="semiBold" />
          </View>
        </View>
      ))}
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  container: {
    flexGrow: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  legendContainer: {
    flexDirection: 'row',
    marginBottom: 16,
    marginHorizontal: 12,
  },
  color: {
    marginEnd: 6,
    marginTop: 4,
    height: 13,
    width: 13,
    borderRadius: 2,
  },
  textColor: {
    color: theme.colors.darkTint4,
  },
});

const memoizedComponent = React.memo(GraphLegends);
export { memoizedComponent as GraphLegends };
