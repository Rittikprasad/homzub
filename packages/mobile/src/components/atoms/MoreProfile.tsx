import React, { Component } from 'react';
import { StyleSheet, TouchableOpacity, View, ViewStyle } from 'react-native';
import { connect } from 'react-redux';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';
import { Avatar } from '@homzhub/common/src/components/molecules/Avatar';
import { UserProfile as UserProfileModel } from '@homzhub/common/src/domain/models/UserProfile';

interface IProps {
  onIconPress: () => void;
  headerContainerStyle?: ViewStyle;
}

interface IStateProps {
  userProfile: UserProfileModel;
}

type Props = IProps & IStateProps;

class MoreProfile extends Component<Props> {
  public render(): React.ReactNode {
    return <View style={styles.container}>{this.renderHeader()}</View>;
  }

  private renderHeader = (): React.ReactElement | null => {
    const { onIconPress, userProfile, headerContainerStyle } = this.props;

    return (
      <TouchableOpacity onPress={onIconPress} style={[styles.headerContainer, headerContainerStyle]}>
        <View style={styles.flexRow}>
          <Avatar
            isOnlyAvatar
            imageSize={60}
            fullName={userProfile?.name ?? 'User'}
            initialsContainerStyle={styles.initialsContainer}
            image={userProfile?.profilePicture ?? ''}
          />
          <View style={styles.nameContainer}>
            <Text maxLength={14} type="regular" textType="semiBold">
              {userProfile?.name ?? 'User'}
            </Text>
            <Label type="large" textType="semiBold" style={styles.progressMsg}>
              {`${userProfile?.profileProgress}% Profile Completed`}
            </Label>
          </View>
        </View>
        <Icon name={icons.rightArrow} size={18} color={theme.colors.lowPriority} />
      </TouchableOpacity>
    );
  };
}

const mapStateToProps = (state: IState): IStateProps => {
  const { getUserProfile } = UserSelector;
  return {
    userProfile: getUserProfile(state),
  };
};

const moreProfile = connect(mapStateToProps)(MoreProfile);
export { moreProfile as MoreProfile };

const styles = StyleSheet.create({
  container: {
    flex: 1,
    borderTopEndRadius: 4,
    borderTopStartRadius: 4,
  },
  flexRow: {
    flexDirection: 'row',
  },
  headerContainer: {
    flex: 1,
    backgroundColor: theme.colors.moreSeparator,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: theme.layout.screenPadding,
    height: 120,
  },
  initialsContainer: {
    elevation: 10,
    shadowColor: theme.colors.shadow,
    shadowRadius: 4,
  },
  nameContainer: {
    marginStart: 12,
    justifyContent: 'center',
  },
  progressMsg: {
    color: theme.colors.green,
  },
});
