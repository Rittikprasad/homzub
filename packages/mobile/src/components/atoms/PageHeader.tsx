import React, { memo, ReactElement } from 'react';
import { StyleSheet, View } from 'react-native';
import { theme } from '@homzhub/common/src/styles/theme';
import { Label, Text, fontLineHeights } from '@homzhub/common/src/components/atoms/Text';

export interface IPageHeaderProps {
  contentTitle?: string;
  contentSubTitle?: string;
  contentLink?: string;
  disableDivider?: boolean;
  onLinkPress?: () => void;
}

const PageHeader = ({
  contentTitle,
  contentSubTitle,
  contentLink,
  disableDivider = false,
  onLinkPress,
}: IPageHeaderProps): ReactElement | null => {
  if (!contentTitle) {
    return null;
  }

  return (
    <View
      style={[
        styles.titleContainer,
        !disableDivider && {
          borderBottomColor: theme.colors.disabled,
          borderBottomWidth: StyleSheet.hairlineWidth,
        },
      ]}
    >
      <Text type="large" textType="semiBold" style={styles.contentTitle}>
        {contentTitle}
      </Text>
      {!!contentSubTitle && (
        <View style={styles.subTitleContainer}>
          <Label type="large" style={styles.contentSubTitle}>
            {contentSubTitle}
          </Label>
          {contentLink && (
            <Label type="large" textType="semiBold" style={styles.linkText} onPress={onLinkPress}>
              {` ${contentLink}`}
            </Label>
          )}
        </View>
      )}
    </View>
  );
};

const memoizedComponent = memo(PageHeader);
export { memoizedComponent as PageHeader };

const PADDING_TOP = 16;
export const TITLE_HEIGHT = PADDING_TOP + fontLineHeights.text.large;

const styles = StyleSheet.create({
  subTitleContainer: {
    paddingTop: 6,
    flexDirection: 'row',
  },
  titleContainer: {
    paddingTop: PADDING_TOP,
    paddingBottom: 12,
    paddingHorizontal: theme.layout.screenPadding,
    backgroundColor: theme.colors.white,
  },
  contentTitle: {
    color: theme.colors.darkTint1,
  },
  contentSubTitle: {
    color: theme.colors.darkTint4,
  },
  linkText: {
    color: theme.colors.primaryColor,
  },
});
