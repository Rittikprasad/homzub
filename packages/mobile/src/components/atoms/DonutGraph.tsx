import React, { useState, useEffect } from 'react';
import { StyleSheet } from 'react-native';
import { PieChart } from 'react-native-svg-charts';
import { isEqual } from 'lodash';
import { FinanceUtils } from '@homzhub/common/src/utils/FinanceUtil';
import { ObjectUtils } from '@homzhub/common/src/utils/ObjectUtils';
import { EmptyState } from '@homzhub/common/src/components/atoms/EmptyState';
import { GraphLegends } from '@homzhub/mobile/src/components/atoms/GraphLegends';
import { GeneralLedgers, IGeneralLedgerGraphData } from '@homzhub/common/src/domain/models/GeneralLedgers';

const INNER_RADIUS = '50%';
const HEIGHT = 225;
interface IProps {
  data: GeneralLedgers[];
}

const DonutGraph = (props: IProps): React.ReactElement => {
  const { data } = props;
  const [donutData, setDonutData] = useState<IGeneralLedgerGraphData[]>([]);

  useEffect(() => {
    const ledgersByCategory = ObjectUtils.groupBy<GeneralLedgers>(data, 'categoryId');
    const transformedData: IGeneralLedgerGraphData[] = [];

    Object.keys(ledgersByCategory).forEach((categoryId) => {
      const ledgers = ledgersByCategory[categoryId];
      const { category } = ledgers[0];
      let { amount } = ledgers[0];

      if (ledgers.length > 1) {
        amount = ledgersByCategory[categoryId].reduce((acc, ledger) => acc + ledger.amount, 0);
      }

      transformedData.push({
        key: Number(categoryId),
        title: category,
        value: amount,
        svg: { fill: FinanceUtils.getGraphColor(category) },
      });
    });

    setDonutData(transformedData);
  }, [data]);

  const render = (): React.ReactElement => {
    if (data.length === 0) return <EmptyState />;

    return (
      <>
        <PieChart style={styles.pieChart} padAngle={0.025} data={donutData} innerRadius={INNER_RADIUS} />
        <GraphLegends data={donutData} />
      </>
    );
  };
  return render();
};

const styles = StyleSheet.create({
  pieChart: {
    flex: 1,
    marginBottom: 16,
    height: HEIGHT,
  },
});

const memoizedComponent = React.memo(DonutGraph, isEqual);
export { memoizedComponent as DonutGraph };
