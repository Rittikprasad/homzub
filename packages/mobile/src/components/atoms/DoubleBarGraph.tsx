import React from 'react';
import { ScrollView, StyleSheet, View } from 'react-native';
import { useSelector } from 'react-redux';
import { BarChart, Grid, XAxis, YAxis } from 'react-native-svg-charts';
import { PathProps } from 'react-native-svg';
import { isEqual } from 'lodash';
import { FinanceUtils, IGraphProps } from '@homzhub/common/src/utils/FinanceUtil';
import { theme } from '@homzhub/common/src/styles/theme';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { EmptyState } from '@homzhub/common/src/components/atoms/EmptyState';
import { GraphLegends } from '@homzhub/mobile/src/components/atoms/GraphLegends';
import { DateRangeType } from '@homzhub/common/src/constants/FinanceOverview';

interface IProps {
  data: IGraphProps;
}

// SVG OPTIONS
const SVG_FONT = { fontSize: 12, fill: theme.colors.darkTint6, fontWeight: '600' as '600' };
const SVG_GRID = { strokeDasharray: [5, 5], stroke: theme.colors.darkTint7, width: '100%' };
// INSET OPTIONS
const VERTICAL_INSET = { top: 8, bottom: 8 };
// VIEWPORT CONSTANTS
const HEIGHT = 255;
const Y_AXIS_WIDTH = 55;

const DoubleBarGraph = (props: IProps): React.ReactElement => {
  const {
    data: { data1, data2, label, type },
  } = props;

  let HORIZONTAL_INSET = { left: 12, right: 12 };
  let WIDTH = theme.viewport.width * 1.3;

  if (type === DateRangeType.Month) {
    WIDTH = label.length <= 4 ? theme.viewport.width - 64 : theme.viewport.width;
    HORIZONTAL_INSET = { left: 24, right: 24 };
  }

  const currency = useSelector(UserSelector.getCurrency);

  const barData = [
    {
      data: data1,
      svg: { fill: theme.colors.expense } as Partial<PathProps>,
    },
    {
      data: data2,
      svg: { fill: theme.colors.income } as Partial<PathProps>,
    },
  ];

  const isData1Empty = data1.every((value) => value === 0);
  const isData2Empty = data2.every((value) => value === 0);
  if (isData1Empty && isData2Empty) {
    return <EmptyState />;
  }
  return (
    <>
      <ScrollView horizontal showsHorizontalScrollIndicator={false}>
        <View>
          <View style={styles.rowContainer}>
            <YAxis
              contentInset={VERTICAL_INSET}
              data={data1.concat(data2)}
              svg={SVG_FONT}
              formatLabel={(value: number): string => `${currency.currencySymbol} ${value}`}
              style={styles.yAxis}
              numberOfTicks={5}
            />
            <BarChart
              contentInset={VERTICAL_INSET}
              spacingInner={0.6}
              style={{ width: WIDTH, height: HEIGHT }}
              data={barData}
              numberOfTicks={5}
            >
              <Grid direction={Grid.Direction.HORIZONTAL} svg={SVG_GRID} />
            </BarChart>
          </View>
          <XAxis
            spacingInner={0.35}
            style={[styles.xAxis, { width: WIDTH }]}
            data={label}
            contentInset={HORIZONTAL_INSET}
            svg={SVG_FONT}
            formatLabel={(_, index: number): string => label[index]}
          />
        </View>
      </ScrollView>
      <GraphLegends data={FinanceUtils.barGraphLegends(data1, data2)} />
    </>
  );
};

const memoizedComponent = React.memo(DoubleBarGraph, isEqual);
export { memoizedComponent as DoubleBarGraph };

const styles = StyleSheet.create({
  rowContainer: {
    flexDirection: 'row',
  },
  xAxis: {
    marginStart: Y_AXIS_WIDTH,
    paddingVertical: 12,
  },
  yAxis: {
    alignItems: 'flex-start',
    width: Y_AXIS_WIDTH,
  },
});
