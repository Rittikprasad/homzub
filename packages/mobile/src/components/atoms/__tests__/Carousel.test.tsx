import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { OnboardingData } from '@homzhub/common/src/mocks/Onboarding';
import { SnapCarousel } from '@homzhub/mobile/src/components/atoms/Carousel';

const mock = jest.fn();

describe('Carousel Atom', () => {
  let component: any;
  let props: any;

  beforeEach(() => {
    props = {
      carouselData: OnboardingData,
      activeIndex: 0,
      carouselItem: OnboardingData[0],
      bubbleRef: mock,
      onSnapItem: mock,
      testID: 'carsl',
    };
    component = shallow(<SnapCarousel {...props} />);
  });

  afterEach(() => jest.clearAllMocks());

  it('should match snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should not call the bubbleref', () => {
    props = {
      ...props,
      carouselItem: mock,
    };
    component = shallow(<SnapCarousel {...props} />);
    component.dive().find('[testID="carsl"]').prop('onLayout')();
    expect(mock).toHaveBeenCalled();
  });
});
