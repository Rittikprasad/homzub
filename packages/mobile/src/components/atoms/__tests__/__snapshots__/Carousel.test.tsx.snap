// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Atom should match snapshot 1`] = `
<Carousel
  activeAnimationOptions={null}
  activeAnimationType="timing"
  activeSlideAlignment="center"
  activeSlideOffset={20}
  apparitionDelay={0}
  autoplay={false}
  autoplayDelay={1000}
  autoplayInterval={3000}
  callbackOffsetMargin={5}
  containerCustomStyle={Object {}}
  contentContainerCustomStyle={Object {}}
  data={
    Array [
      Object {
        "description": "Sit amet, in enim nunc, etiam amet velit. Malesuada non neque nec in a.",
        "image_url": "https://homzhub-bucket.s3.amazonaws.com/Search_Onboarding.svg",
        "title": "Search",
      },
      Object {
        "description": "Sit amet, in enim nunc, etiam amet velit. Malesuada non neque nec in a.",
        "image_url": "https://homzhub-bucket.s3.amazonaws.com/Post_Onboarding.svg",
        "title": "Post",
      },
      Object {
        "description": "Sit amet, in enim nunc, etiam amet velit. Malesuada non neque nec in a.",
        "image_url": "https://homzhub-bucket.s3.amazonaws.com/Maintain_Onboarding.svg",
        "title": "Maintain",
      },
    ]
  }
  enableMomentum={false}
  enableSnap={true}
  firstItem={0}
  hasParallaxImages={false}
  inactiveSlideOpacity={0.7}
  inactiveSlideScale={0.9}
  inactiveSlideShift={0}
  initialNumToRender={10}
  itemWidth={720}
  layout="default"
  lockScrollTimeoutDuration={1000}
  lockScrollWhileSnapping={false}
  loop={false}
  loopClonesPerSide={3}
  onEndReached={[Function]}
  onLayout={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEnabled={true}
  shouldOptimizeUpdates={true}
  slideStyle={Object {}}
  sliderWidth={750}
  swipeThreshold={20}
  testID="carsl"
  useScrollView={false}
  vertical={false}
/>
`;
